### PR TITLE
allow multiple attributes per style-spec property

### PR DESCRIPTION
**Pull request by mollymerp**
This refactor is necessary for #2732 because `*-pattern` properties are cross-faded and require more data to be passed to the shaders – the x,y coordinates of the top-left and bottom-right corners of up to four icons in an image_atlas, and the pixel ratio of the atlas – that can be packed in one `vec4` attribute.

I added a function that takes a `property` and `type` and returns the `PaintVertexArray` constructor for the layout needed for that property/type/expression type. This is where we will be able to add exceptions to the default paint vertex array layouts for properties with multiple attributes. 

I haven't settled on a good way to extend the `populatePaintArrays` functionality to accommodate special cases yet, but it might make sense to work that out in the follow up PR implementing `line-pattern` property functions. 

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [ ] write tests for all new functionality
 - [ ] document any changes to public APIs
 - [ ] post benchmark scores
 - [x] manually test the debug page

### CODE REVIEW

File: src/data/program_configuration.js [hunk 1]
```diff
@@ -68,19 +71,19 @@ interface Binder<T> {
...
         this.value = value;
-        this.name = name;
+        this.names = names;
         this.type = type;
         this.statistics = { max: -Infinity };
     }
 
     defines() {
-        return [`#define HAS_UNIFORM_u_${this.name}`];
+        return this.names.map((name) => { return `#define HAS_UNIFORM_u_${name}`; });
```
Comment (mourner) [inline src/data/program_configuration.js]: Nit: you can omit braces and `return` here.

File: src/data/program_configuration.js [hunk 2]
```diff
@@ -93,36 +96,40 @@ class ConstantBinder<T> implements Binder<T> {
...
-        } else {
-            gl.uniform1f(program.uniforms[`u_${this.name}`], value);
-        }
+        this.names.forEach((name)=>{
+            if (this.type === 'color') {
+                gl.uniform4f(program.uniforms[`u_${name}`], value.r, value.g, value.b, value.a);
+            } else {
+                gl.uniform1f(program.uniforms[`u_${name}`], value);
+            }
+        });
```
Comment (mourner) [inline src/data/program_configuration.js]: Nit: let's avoid using `forEach` and closures in favor of a `for .. of` loop.

File: src/data/program_configuration.js [hunk 3]
```diff
@@ -93,36 +96,40 @@ class ConstantBinder<T> implements Binder<T> {
...
-        }];
+        const PaintVertexArray = layout;
+        this.paintVertexAttributes = names.map((name)=> {
+            return {
+                name: `a_${name}`,
+                type: 'Float32',
+                components: type === 'color' ? 2 : 1,
+                offset: 0
+            };
+        });
```
Comment (mourner) [inline src/data/program_configuration.js]: Nit: you can omit braces and `return` here too if you surround the object with `(...)`.

**Comment by mollymerp** [comment]
folding this into #6289 – I incorporated your requested changes @mourner 😸

**Pull request closed without merging**